src/water_cmd.cpp

```diff
@@ -478,7 +478,7 @@ CommandCost CmdBuildCanal(TileIndex tile, DoCommandFlag flags, uint32 p1, uint32
 		bool water = IsWaterTile(current_tile);
 
 		/* can't make water of water! */
-		if (water && (!IsTileOwner(current_tile, OWNER_WATER) || wc == WATER_CLASS_SEA)) continue;
+		if (water && IsCanal(current_tile) && Company::IsValidID(_current_company) && (IsTileOwner(current_tile, _current_company) || GetTileOwner(current_tile) == OWNER_NONE)) continue;
```

Comment:
As mentioned on IRC, not explaining what this does makes this a terrible line to read. So many questions:

- Why the `IsValidID`?
- Why `IsTileOwner` vs `GetTileOwner`?
- Why the check against `_current_company`?
- Why the check against `OWNER_NONE`?
- Where did the check against `OWNER_WATER` go?
- Where did the `WATER_CLASS_SEA` check go?

These might all be legit choices and be a fine solution, but nothing in this PR nor description helps me at all. You force me to read up in this code in detail to figure it out .. by that time it is often easier for me to just do it myself, instead of trying to understand what you tried to do.

Please please do better. Fix the comment above the line, and try to explain what you think this line is doing. We will take it from there. But this really is not sufficient, sorry :(

Reply:
I used `Company::IsValidID(_current_company)` because in SE, you're `OWNER_NONE` and in a game `_current_company` is a valid company. I needed to make the check fail in SE to allow overbuilding canals on SE-placed canals. Not sure if using `_current_company != OWNER_NONE` instead would be better - In-game I need the entire `if` to be true to skip the `current_tile` to come up with "already built" message. In the case of building a company canal over a canal of another company, I need `Company::IsValidID(_current_company)` to fail.

The check `GetTileOwner(current_tile) == OWNER_NONE` is only happening in-game, and not in SE. This is made this way to avoid overbuilding canals on SE-built/bankrupted canals, but mostly because it'd be paying the price for something that already exists. I though this behaviour was fine and didn't need change, I kept it the way it was already. This is in contrast with what happens in scenario editor, where I made it overbuild.

The check against `OWNER_WATER` and `WATER_CLASS_SEA ` are gone. Trying to explain why is actually tougher than I expected... it's mostly because of the erroneous messages that show up especially on SE. If I ignore the messages for a moment, the logic still seems valid at glance: overbuilding sea on sea doesn't have to clear the tile just to build over again. This PR makes it do the opposite, it clears the tile and builds over with the benefit that the message is gone. The drawback is that the cost is added, but since it's in SE, it doesn't really matter (or does it?).


As for what should the comment in the line above should say? I'm not really sure, honestly. It can make water of water now in some cases, with a few exceptions related to whenever the tile is a canal and who owns it. I'm not really sure what to write in the comment.